src/HealthChecks.ApplicationStatus/DependencyInjection/ApplicationStatusHealthCheckBuilderExtensions.cs

```diff
@@ -0,0 +1,72 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.HealthChecks;
+using Microsoft.Extensions.Hosting;
+
+namespace HealthChecks.ApplicationStatus.DependencyInjection;
+public static class ApplicationStatusHealthCheckBuilderExtensions
+{
+    private const string NAME = "applicationstatus";
+
+    /// <summary>
+    /// Add a health check for Application Status.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHealthChecksBuilder"/>.</param>
+    /// <param name="lifetime">Represents an object for intercepting application lifetime events.</param>
+    /// <param name="name">The health check name. Optional. If <c>null</c> the type name 'applicationstatus' will be used for the name.</param>
+    /// <param name="failureStatus">
+    /// The <see cref="HealthStatus"/> that should be reported when the health check fails. Optional. If <c>null</c> then
+    /// the default status of <see cref="HealthStatus.Unhealthy"/> will be reported.
+    /// </param>
+    /// <param name="tags">A list of tags that can be used to filter sets of health checks. Optional.</param>
+    /// <param name="timeout">An optional <see cref="TimeSpan"/> representing the timeout of the check.</param>
+    /// <returns>The specified <paramref name="builder"/>.</returns>
+    public static IHealthChecksBuilder AddApplicationStatus(
+        this IHealthChecksBuilder builder,
+        IHostApplicationLifetime lifetime,
+        string? name = default,
+        HealthStatus? failureStatus = default,
+        IEnumerable<string>? tags = default,
+        TimeSpan? timeout = default)
+    {
+        if (lifetime == null)
+        {
+            throw new ArgumentNullException(nameof(IHostApplicationLifetime));
+        }
+
+        return builder.Add(new HealthCheckRegistration(
+            name ?? NAME,
+            sp => new ApplicationStatusHealthCheck(lifetime),
```

Comment:
You introduced memory leak in both methods since health check is created each time someome calls health checks API. You call `Register` in healthcheck ctor over and over.

Reply:
Don't get this one. I follow examples from other healthchecks. Should I add a validation so there cannot be two ApplicationStatus healthchecks?

Reply:
They can if you are talking about extension method that register HC. I'm talking about something else. Just put a breakpoint into `ApplicationStatusHealthCheck` constructor, start your test app and make some requests to touch you /health or whatever URL. You will see that **each** web request yields a call to ctor. You programmed the HC registration in such a way. Nothing criminal here yet. Yes, you may find other checks that do the same. But... criminal is in line` _lifetime.ApplicationStopping.Register(OnStopping);` You register delegate into cancellation token of "application-wide" object (IHostApplicationLifetime)  **each** time web request come. Where do you unregister this delegate? Nowhere. So registration list will grow over and over. How to overcome this? For example pull HC from DI. See `NatsHealthCheckBuilderExtensions` for example. Note that it implements `IDisposable`. You **may** also implement it and unregister delegate.

Reply:
Yes, you are right. This delegate registration is a killer. I will fix it

Reply:
So I changed the way how it is registered. At this point, you can add as many health checks as you want to the builder. Still, there will be only one instance of `ApplicationStatusHealthCheck` because it is fetched from IoC directly without creating a new instance.